File: 785-is-graph-bipartite/785-is-graph-bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        
        int V = graph.size();
        vector<int>vis(V,0);
         queue<pair<int,int>>qu;
         for(int i = 0 ; i < V ; i++){
               if(vis[i] != 0)
                   continue;
              
               qu.push({i,-1});
               qu.push({-1,-1});
               int color = 1;
             
               while(qu.size() > 0 ){
                   
                   pair<int,int> top = qu.front();
                   qu.pop();
                   
                   if(top.first == -1){
                       if(qu.size() > 0 )
                           qu.push({-1,-1});
                       
                       color = 3 - color;
                       
                   }else{
                       
                       vis[top.first] = color;
                   
                    for(int j : graph[top.first]){
                       
                       if(j == top.second)
                           continue;
                       if(vis[j] == vis[top.first])
                           return false;
                       if(vis[j] == 0){
                           qu.push({j,top.first});
                       }
                     }
                       
                   }
                   
                   
               }
          
         }
        return true;
    }
};

File: 785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int V = graph.length;
        int[] vis = new int[V];
        
// vis will have 3 states : //  0 - not visited //  1 - visited and coloured with 1 //2 - visited and coloured with 2
        
        for(int i = 0 ; i < V ; i++){ 
            // loop used as graph may bedisconnected so we need to check for all nodes na re babu...
            if(vis[i] == 0 && dfs(graph,vis,i,-1,1) == false)
                 return false;   
        }
        return true;
    }
    
    public boolean dfs(int[][] graph,int[] vis ,int curr,int par , int color ){
        
        vis[curr] = color;
        
        for(int j : graph[curr]){
              if(j == par) // respect your parent 
                 continue; 
            
              if(vis[j] == color)
                  return false;
              if(vis[j] == 0 && dfs(graph,vis,j,curr,3-color) == false)
                  return false;
        }
        return true;
        
        
    }
    
    
}